src/traces/carpet/axis_defaults.js

```diff
@@ -0,0 +1,238 @@
+ *  noHover: boolean, this axis doesn't support hover effects?
+ *  data: the plot data to use in choosing auto type
+ *  bgColor: the plot background color, to calculate default gridline colors
+ */
+module.exports = function handleAxisDefaults(containerIn, containerOut, options) {
+    var letter = options.letter,
+        font = options.font || {},
+        attributes = carpetAttrs[letter + 'axis'];
+
+    options.noHover = true;
+
+    function coerce(attr, dflt) {
+        return Lib.coerce(containerIn, containerOut, attributes, attr, dflt);
+    }
+
+    // I don't know what this does:
+    function coerce2(attr, dflt) {
+        return Lib.coerce2(containerIn, containerOut, attributes, attr, dflt);
+    }
+
+    // set up some private properties
+    if(options.name) {
+        containerOut._name = options.name;
+        containerOut._id = options.name;
+    }
+
+    // now figure out type and do some more initialization
+    var axType = coerce('type');
+    if(axType === '-') {
+        if(options.data) setAutoType(containerOut, options.data);
+
+        if(containerOut.type === '-') {
+            containerOut.type = 'linear';
+        }
+        else {
+            // copy autoType back to input axis
+            // note that if this object didn't exist
+            // in the input layout, we have to put it in
+            // this happens in the main supplyDefaults function
+            axType = containerIn.type = containerOut.type;
+        }
+    }
+
+    coerce('smoothing');
+    coerce('cheatertype');
+
+    coerce('showticklabels');
+    coerce('labelprefix', letter + ' = ');
+    coerce('labelsuffix');
+    coerce('showtickprefix');
+    coerce('showticksuffix');
+
+    coerce('separatethousands');
+    coerce('tickformat');
+    coerce('exponentformat');
+    coerce('showexponent');
+    coerce('categoryorder');
+
+    coerce('tickmode');
+    coerce('tickvals');
+    coerce('ticktext');
+    coerce('tick0');
+    coerce('dtick');
+
+    if(containerOut.tickmode === 'array') {
+        coerce('arraytick0');
+        coerce('arraydtick');
+    }
+
+    coerce('labelpadding');
+
+    containerOut._hovertitle = letter;
+
+
+    if(axType === 'date') {
+        var handleCalendarDefaults = Registry.getComponentMethod('calendars', 'handleDefaults');
+        handleCalendarDefaults(containerIn, containerOut, 'calendar', options.calendar);
+    }
+
+    setConvert(containerOut, options.fullLayout);
+
+    var dfltColor = coerce('color', options.dfltColor);
+    // if axis.color was provided, use it for fonts too; otherwise,
+    // inherit from global font color in case that was provided.
+    var dfltFontColor = (dfltColor === containerIn.color) ? dfltColor : font.color;
+
+    coerce('title');
+    Lib.coerceFont(coerce, 'titlefont', {
+        family: font.family,
+        size: Math.round(font.size * 1.2),
+        color: dfltFontColor
+    });
+
+    coerce('titleoffset');
+
+    coerce('tickangle');
+
+    var validRange = (
```

Comment:
you can now use [`ax.isValidRange`](https://github.com/plotly/plotly.js/blob/master/src/plots/cartesian/set_convert.js#L406). :recycle: